[AArch64][SVE2] Do not emit RSHRNB for large shifts

MDevereau opened the pull request:
rshrnb's shift amount operand must be between 1-EltSizeInBits. This patch stops RSHRNB ISD nodes being emitted in this case

MDevereau committed:
[AArch64][SVE2] Do not emit RSHRNB for large shifts

rshrnb's shift amount operand must be between 1-EltSizeInBits.
This patch stops RSHRNB ISD nodes being emitted in this case

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20241,6 +20241,9 @@ static SDValue trySimplifySrlAddToRshrnb(SDValue Srl, SelectionDAG &DAG,
     return SDValue();
   unsigned ShiftValue = SrlOp1->getZExtValue();
 
+  if (ShiftValue > ResVT.getScalarSizeInBits())
```
Is there a reason the shift amount cannot be clamped to the maximum rather than not doing the transformation?

MDevereau replied:
@paulwalker-arm I was trying to think of a case but since the add is check for single use and the top half of the elements get narrowed I think this it's safe to clamp it to the maximum value.

MDevereau committed:
[AArch64][SVE2] Do not emit RSHRNB for large shifts

rshrnb's shift amount operand must be between 1-EltSizeInBits.
This patch stops RSHRNB ISD nodes being emitted in this case

david-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20241,6 +20241,10 @@ static SDValue trySimplifySrlAddToRshrnb(SDValue Srl, SelectionDAG &DAG,
   unsigned ShiftValue = SrlOp1->getZExtValue();
 
+  uint64_t EltSize = ResVT.getScalarSizeInBits();
+  if (ShiftValue > EltSize)
```
I discussed this with @paulwalker-arm offline and we don't think this is right after all. I think we should just bail out of the optimisation if the ShiftValue exceeds the truncated element width. Also, it would be good to at least add an assert here that the shift value is zero. From what you're saying it should never happen, which makes sense.

MDevereau replied:
Can you please explain why you both came to the conclusion that this isn't correct?

paulwalker-arm replied:
Sorry Matt this was my fault. My question arose from the typically behaviour of right shifts that are greater than the element bit length, whereby `i32 >> #32+N == i32 >> #32`.

Dave pointed out that in this case the element type in question is actually half sized because the combine is only expected to be called when a truncate is in play. That would imply `i32 >> 16+N == i32 >> 16`, which is clearly bogus.

MDevereau replied:
Thanks. I've now just gone with
```c++
  if (ShiftValue < 1 || ShiftValue > ResVT.getScalarSizeInBits())
    return SDValue();
```

MDevereau committed:
[AArch64][SVE2] Do not emit RSHRNB for large shifts

rshrnb's shift amount operand must be between 1-EltSizeInBits.
This patch stops RSHRNB ISD nodes being emitted in this case

david-arm reviewed (approved):
LGTM! Thanks @MDevereau.
